### ISSUE TITLE
Request handled by wrong route

### ISSUE THREAD

**Issue by elyrank** [softwaremill/tapir#92]
when defining 2 routes: one with /xx/{id} (id defined as Long) and another with /xx/yy
the request to /xx/yy is handled by the route for /xx/{id}

the error I get for it is: Invalid value for: path
which is not very informative, I would rather also get the caused by - as only by debugging it I understood what was the bug.
in my case that would have been: numberFormatException on "export" 

I have the following base endpoints: (they contain previous parts of the paths - 3 Ints as path param)
```
val notionEndpoint = baseEndpoint.in("notion")
val notionIdEndpoint = notionEndpoint.in(path[Long]("notionId"))
```

then I have some routes:
```
 val getNotion: ServerEndpoint[(Int, Int, Int, Long), Unit, Option[Notion], Nothing, Future] =
      notionIdEndpoint.get.summary("Get a single notion").tag("Notion")
        .out(jsonBody[Option[Notion]])
        .serverLogic((getNotionLogic _).tupled)

val export: ServerEndpoint[(Int, Int, Int, Boolean, Boolean), Unit, (String, Nodes), Nothing, Future] = {
      notionEndpoint.get
        .in("export")
        .in(query[Option[Boolean]]("includeX")
          .example(Some(true))
          .map(_.getOrElse(true))(Option.apply))
        .in(query[Option[Boolean]]("includeY")
          .example(Some(true))
          .map(_.getOrElse(true))(Option.apply))
        .out(header[String]("Content-Disposition"))
        .out(jsonBody[Nodes])
        .tag("Import Export")
        .serverLogic((exportLogic _).tupled)
    }
```

and lastly - I return a list of endpoints to be later transformed to routes:
```
List(getNotion, export)
```

workaround: manually set the order of the endpoints for the export to be first

**Comment by adamw** [comment]
There's two things in tapir which might help with such problems; it's possible that they need either better documentation or better defaults - please let me know what's your opinion :)

First of all, there's the DEBUG-level logging: https://tapir-scala.readthedocs.io/en/latest/server/debugging.html. By default, there's a DEBUG-level log when an endpoint handles a request - so that you can identify which endpoint is called. You can also optionally turn on logging of all decode failures - to find out, which endpoints reject the request and why.

Secondly, you can customise the errors by providing a custom `DecodeFailureHandler`, see `ServerDefaults`. Both the failing input & failure details are available. I would be reluctant to include anything from the exception when there's a decode failure (as it might expose implementation details), but again, maybe the defaults might be better.

**Comment by elyrank** [comment]
I looked at the docs now, and indeed it gave all the info that was needed.
the defaults are good enough ( BTW - there are only 3 options, not 4 as in the docs)
I just needed to figure out how to enable it. would be useful to add it to the doc

in reference.ccnf
```
akka {
  loglevel = DEBUG
}
```

but this is not the main issue - it is still handled by the wrong route.

**Comment by elyrank** [comment]
BTW - I think it started from one of the latest versions

**Comment by adamw** [comment]
Thanks, I fixed the docs :)

So the main issue is that you have two endpoints:
1. with path `/xx/yy`
2. with path `/xx/{id}`
and the wrong one handle the request?

Is `yy` a constant? How do you order the endpoints? I think I that if the endpoints with the constant values are defined first, and then the wildcard endpoints (which capture any value) are listed second, things should work. 

I would also expect an error if the order is different.

**Comment by elyrank** [comment]
when the order is reversed  -i.e. when {id} route is first - there is an error:
` Invalid value for: path`
but if you would define it in akka-http you would use it like this:
```pathPrefix("xx" / LongNumber) { id =>```
which will catch only Longs and not Strings

**Comment by adamw** [comment]
I see. I'm not sure what the default behavior should be here.

If you would have:

1. `"api" / path[String]`
2. `"api" / "xyz"`

then the all requests would be captured by the first endpoint - as it's a "wildcard" endpoint. And I suppose there's no surprises here.

Now we change the `String` to a `Long` - so there's a required format that the path must match. Now parsing the path might result in a `NumberFormatException`. Should such an exception be ignored? 

In the end, it boils down to this line in the default decode failure handler:

```
case _: EndpointInput.PathCapture[_] if failure.isInstanceOf[DecodeResult.Error] =>
          response(StatusCodes.BadRequest, s"Invalid value for: path")
```

without it, the result would be `DecodeFailureHandling.noMatch` and the next endpoint would be tried.

**Comment by elyrank** [comment]
I think the best way is to remove that line and return a no match.

**Comment by adamw** [comment]
@elyrank agreed :) I think https://github.com/softwaremill/tapir/commit/a3b61ab7180fd2c5c13745ae5602b0f6e017e903 should solve the problem?

**Comment by elyrank** [comment]
Yep :)

**Comment by adamw** [comment]
Released in 0.7.10
